### PR TITLE
ISPN-6381 - Add more tests for js-client

### PR DESCRIPTION
**Pull request by andyuk1986**
The CRUD and entry expiration tests are also ported from local mode to clustered mode. 
The infinispan_local_spec.js is changed so that the tests are run for both available Hotrod Protocol Versions. 
More tests are added verifying the codec.js and funcitonal.js library functions. 

P.S. There are a few tests in infinispan_cluster_spec.js which are disabled due to the bugs reported in the JIRA.

### CODE REVIEW

**Comment by galderz** [comment]
@andyuk1986 Thanks a lot for adding all these tests! I'll be looking into them today 😃

File: lib/codec.js [hunk 1]
```diff
@@ -18,6 +18,7 @@
   exports.encodeBytes = f.lift(doEncodeBytes, _.identity);
   exports.encodeString = f.lift(doEncodeString, _.identity);
   exports.encodeSignedInt = f.lift(doEncodeSignedInt, _.identity);
+  exports.encodeShort = f.lift(doEncodeShort, _.identity);
```
Comment (galderz) [inline lib/codec.js]: Is there a need for this? As far as I know, the protocol does not currently encode any shorts.

File: spec/infinispan_cluster_spec.js [hunk 2]
```diff
@@ -30,11 +33,467 @@ describe('Infinispan cluster client', function() {
       .catch(t.failed(done)).finally(done);
   });
 
+  it('can put a key/value pair using different clients', function(done) {
```
Comment (galderz) [inline spec/infinispan_cluster_spec.js]: These tests feel a bit repetitive compare to what is done in local spec. There's an opportunity here to refactor the local spec tests so that you can pass in two clients: the client that does the operation (e.g. put) and the client that checks the result (e.g. get), and then call it different clients. In the case of local, both the client that does the operation and checks the result is the same.

Also, I'm not sure the tests you've written here are doing what you expect them to do. What is it that you are trying to verify here to be precise? 

The reason I ask is because no matter which cluster node each client connects to, it will get the topology and it will route request to the owner node (they are talking to a distributed cache), hence all you are testing is that if a client 1 goes to node A to put a value, client 2 will go to node A to retrieve the value.

I am wondering whether you real intention was to have client 1 store data in node A and then client 2 being able to retrieve the value from node B? If that was your intention, this test is not really achieving that.

File: spec/infinispan_cluster_spec.js [hunk 3]
```diff
@@ -30,11 +33,467 @@ describe('Infinispan cluster client', function() {
...
+        //   return client3.then(t.seqIterator(3, pairs)) // Iterate all data, 3 elements at time, sequential
+        //       .then(function() {
+        //         return client;
+        //       });
+        // })
+        .catch(t.failed(done))
+        .finally(done);
+  });
+
+  xit('can iterate over entries getting their expirable metadata', function(done) {
```
Comment (galderz) [inline spec/infinispan_cluster_spec.js]: Test name should include "using different clients" in the end?

File: spec/infinispan_cluster_spec.js [hunk 4]
```diff
@@ -30,11 +33,467 @@ describe('Infinispan cluster client', function() {
...
+        .catch(t.failed(done)).finally(done);
+  });
+
   // Since Jasmine 1.3 does not have afterAll callback, this disconnect test must be last
   it('disconnects client', function(done) { client
       .then(t.disconnect())
       .catch(t.failed(done))
       .finally(done);
+
+    client2
```
Comment (galderz) [inline spec/infinispan_cluster_spec.js]: If the first client fails to disconnect this won't be invoked and neither client3's disconnect method. Instead, this should be written in such way that you disconnect first client, then disconnect second client... and a single catch + finally section.

File: spec/infinispan_local_spec.js [hunk 5]
```diff
@@ -5,370 +5,274 @@ var readFile = Promise.denodeify(require('fs').readFile);
 var f = require('../lib/functional');
 var t = require('./utils/testing'); // Testing dependency
 
-describe('Infinispan local client', function() {
-  var client = t.client(t.local);
+function describeInfinispanBehaviourInLocalModeWithDifferentHotrodProtocol(options) {
+  describe('with hotrod protocol: ' + options.version, function () {
+    var client = t.client(t.local, null, options.version);
 
-  beforeEach(function(done) { client
```
Comment (galderz) [inline spec/infinispan_local_spec.js]: This section seems to only contain cosmetic changes: moving initial client variable to the next line. Any particular reason for doing this? In any case, if you want to make formatting changes add them as different commits.

**Comment by galderz** [comment]
A few comments, mostly related to repetition or intention. Just being thorough with code changes :) 

Thanks a lot for the hard work put into this! :D

**Comment by galderz** [comment]
I've verified the tests too and they all pass locally :)

I noticed how you disabled some of them since they don't yet pass. Nice trick :D

File: lib/codec.js (hunk 1 above)
Comment (andyuk1986) [inline lib/codec.js]: Ok, got it. Actually I have added this for being able to add tests for encodeShort as well as for increasing the code coverage for codec.js. But if this will be removed in the future, then I will remove this with corresponding tests.

File: spec/infinispan_cluster_spec.js (hunk 2 above)
Comment (andyuk1986) [inline spec/infinispan_cluster_spec.js]: Hmmmm, I see - I haven't taken it into account. 
Yes, you are completely right - that was the purpose of these tests. Then I guess the easiest way would be to run these tests on replicated cache or maybe enable the L1 on distributed cache. What do you think? 

For the refactoring - great idea! I will change that.

File: spec/infinispan_local_spec.js (hunk 5 above)
Comment (andyuk1986) [inline spec/infinispan_local_spec.js]: Yeah I agree @galderz . I was thinking about that as well.
Well the simplest solution would be to set some environment variable with protocol version, and then use it in the spec with process.env.VAR_NAME  reference. 
I will look closer into the arguments provided by jasmine-node to find out if it has anything which I can apply.

Comment (andyuk1986) [inline spec/infinispan_local_spec.js]: @galderz jasmine-node has an argument - 
--config NAME VALUE- set a global variable in process.env

So  I can change the specs so that they are looking for process.env.VAR variable. If it exists then all specs will run with the same protocol version, otherwise 2.2 will be used as default. WDYT?

File: spec/infinispan_cluster_spec.js (hunk 2 above)
Comment (galderz) [inline spec/infinispan_cluster_spec.js]: Actually, regardless of whether it's distributed or replicated, if the client is hash topology aware it'll get hash topology for both so it'll route it according to that.

It's interesting to understand that at the client level, what's interesting is how it does the routing. That's why executing iterator against a cluster of servers is interesting because the way it should route is that it should be sticky. So, whichever node where the iterator gets started, it should direct the rest of reqs for that node to the same node. However, the current code does round robin so that's why iterator w/ cluster fails ([JDG-184](https://issues.jboss.org/browse/JDG-184)).

By explaining this what I'm trying to get to is that it's not wise to add tests for the shake of it unless their really testing something that will highlight client issues. That's why I'm not 100% sure having multi-client tests will add a lot to the testsuite. I know there are some tests that fail in your PR, but I need to inspect them closely to see if they can be replicated with a single client because each client created should be independent (each has their own independent components).

IOW, if you can replicate the same issue with 1 client, it's easier to debug/reproduce so we should strive to do that. If there really are tests that fail with multiple clients, we should look and understand what's the reason behind it and then decide to add more such tests.

Having said that, I'm happy to integrate the tests you have so far with multiple clients, but might refactor them later once I've spotted the root cause.

File: spec/infinispan_local_spec.js (hunk 5 above)
Comment (galderz) [inline spec/infinispan_local_spec.js]: +1 to have an environment variable to drive testing particular protocol version.

> I have played a bit with this argument, and seems that after running the testsuite with e.g. --config A B, the process.env.A = B . Then the testsuite is run without specifying the config, still the process.env.A = B. I couldn't find a way of unsetting it though, if only passing another value for A while running the test suite or deleting it manually after the testsuite is finished (delete process.env.A;) -> not sure how to catch this event.

Ok, if we can't get around that limitation, we can always have that as a CI or test configuration. So when you test it, you'd execute as a two step process:

    $ jasmine-node spec // <- should run with protocol 2.5
    $ jasmine-node spec --config protocol 2.2

The only thing to remember is that some tests won't work with older versions, so we'd have to exclude those somehow.

File: lib/codec.js (hunk 1 above)
Comment (galderz) [inline lib/codec.js]: To provide a bit more information on this, it appears that the root cause in [JDG-94](https://issues.jboss.org/browse/JDG-94) is how JBoss Strings are encoded when adding the script. I have the feeling is that if the script is longer than certain size it fails with the error in the issue. So, what's relevant here to test is how a particular String of certain size is converted into a JBoss Marshalling-encoded String. This is more relevant than simply testing how shorts are encoded on their own. In fact, by testing Strings of different sizes you indirectly test the short encoding and encoding JBoss Strings is something already exposed to the rest of components in the project, so no need to overexpose anything.

Comment (andyuk1986) [inline lib/codec.js]: @galderz thank you for the explanation. You are right - I have added a comment based on my findings in [ISPN-6407](https://issues.jboss.org/browse/ISPN-6407). I have seen that you have already added tests in your upcoming pull request for testing the jboss String conversion.

File: spec/infinispan_cluster_spec.js (hunk 2 above)
Comment (andyuk1986) [inline spec/infinispan_cluster_spec.js]: OK, understood :) Thanks a lot for detailed explanation. So then, I have removed the multi-client tests as there is not purpose in them. I have removed the tests for iterators as well because I have seen that you will add them in your pull request for fix of iterators. 
I have kept the tests for the listeners but only with one client and actually at the moment it would be a duplicate code as we have the same for the local tests. The issues [JDG-186](https://issues.jboss.org/browse/JDG-186) and [JDG-187](https://issues.jboss.org/browse/JDG-187) still appear with one client in clustered mode. I think the JDG-186 appears again because of round robin - the listener is tried to be removed from the other node rather than one where the listener was added. 

As soon as your PRs will be integrated (the ones for numbers returning as well as the iterators fix), I'll rebase this PR, as well as will get rid of duplicate listener tests, and then it can be ready for integration.

File: spec/infinispan_local_spec.js (hunk 5 above)
Comment (andyuk1986) [inline spec/infinispan_local_spec.js]: Implemented this as we have agreed. Also placed a check for iterator tests so that if the protocol is not 2.5, then they are not run.

**Comment by galderz** [comment]
Looks good @andyuk1986. The client disconnect calls in the expiry test is not correct but I can fix that easily. I'll go ahead push this and sort that out. Thanks a lot!
